File: src/ocr_general_doc.py
"""
Gemini APIを使用してPDFファイルのOCR（文字認識）を行い、Markdownデータを出力するプログラム。
一般文書（証拠書類、書籍など）を対象としています。
PDFの各ページを抽出し、Geminiに送信してテキストを抽出します。
コマンドラインから実行し、ファイルまたはディレクトリを指定してまとめて処理できます。

Usage:
    python ocr_general_doc.py <PDF_PATH_OR_DIR> [--batch_size <NUMBER>] [--start_page <NUMBER>] [--end_page <NUMBER>]

出力は、各PDFと同じディレクトリに .md 拡張子で保存されます。
"""

import sys
import argparse
from pathlib import Path

try:
    from lib.pdf_to_markdown import pdf_to_text, get_ocr_prompt
except ImportError:
    # For when running from a different directory or as a module
    try:
        from .lib.pdf_to_markdown import pdf_to_text, get_ocr_prompt
    except ImportError:
        # Fallback if running directly from src
        import sys
        sys.path.append(str(Path(__file__).resolve().parent))
        from lib.pdf_to_markdown import pdf_to_text, get_ocr_prompt

# 一般文書のスタイル定義
GENERAL_DOC_STYLE = """
# CONTEXT: General Document (Evidence, Books, Reports, etc.)
- **Format**: Maintain the original structure as much as possible.
- **Line Breaks**: Merge lines within the same paragraph. Keep line breaks for headings, lists, and clear paragraph transitions.
- **Tables**: If tables are present, represent them using Markdown table format.
- **Lists**: Use standard Markdown list markers (`-`, `*`, or `1.`).
- **Exclusions**: Ignore headers, footers, and page numbers if they are repetitive and not part of the main content.
- **Page Numbers**: Page numbers may be Arabic (1, 2), Kanji (一, 二), or Roman (I, II) numerals. Omit them if they are in margins, but use them for the Page Markers (converting to Arabic numerals).
- **Emphasis**: Use bold (`**text**`) or italics (`*text*`) where appropriate based on the visual style of the document.
"""

if __name__ == "__main__":
    # コマンドライン引数の解析
    parser = argparse.ArgumentParser(description="Gemini Batch APIを使用して一般文書のPDFのOCRを行い、Markdownを出力します。")
    parser.add_argument("input_path", type=str, nargs='?', help="処理対象のPDFファイルパスまたはディレクトリパス")
    parser.add_argument("--batch_size", type=int, default=4, help="一度に処理するページ数 (デフォルト: 4)")
    parser.add_argument("--start_page", type=int, default=1, help="開始ページ番号 (1開始, デフォルト: 1)")
    parser.add_argument("--end_page", type=int, default=None, help="終了ページ番号 (1開始, デフォルト: 最終ページ)")
    parser.add_argument("--show_prompt", action="store_true", help="実行せずにプロンプトのみを表示します。")
    
    args = parser.parse_args()

    if args.show_prompt:
        print("\n--- Gemini OCR Prompt Template ---")
        print(get_ocr_prompt(args.batch_size, GENERAL_DOC_STYLE))
        print("----------------------------------\n")
        sys.exit(0)

    if not args.input_path:
        parser.error("the following arguments are required: input_path (unless --show_prompt is used)")

    input_path = Path(args.input_path)
    
    if input_path.is_dir():
        # ディレクトリ内の全てのPDFファイルを対象にする（ファイル名順にソート）
        pdf_files = sorted(list(input_path.glob("*.pdf")), key=lambda x: x.name)
        if not pdf_files:
            print(f"[WARN] No PDF files found in directory: {input_path}")
            sys.exit(0)
        
        print(f"[INFO] Found {len(pdf_files)} PDF files in {input_path}")
        for pdf_file in pdf_files:
            print(f"\n[PROCESS] Starting: {pdf_file.name}")
            pdf_to_text(
                pdf_file, 
                batch_size=args.batch_size, 
                start_page=args.start_page, 
                end_page=args.end_page,
                context_instruction=GENERAL_DOC_STYLE
            )
    else:
        # 単一ファイルの処理
        pdf_to_text(
            input_path, 
            batch_size=args.batch_size, 
            start_page=args.start_page, 
            end_page=args.end_page,
            context_instruction=GENERAL_DOC_STYLE
        )

File: src/ocr_general_doc.js
/**
 * Gemini APIを使用して一般文書PDFのOCRを行い、Markdownを出力するプログラム。
 * 
 * 使い方:
 *   node src/ocr_general_doc.js <PDFファイルパス または ディレクトリパス> [--batch_size <枚数>] [--start_page <開始ページ>] [--end_page <終了ページ>]
 */
const fs = require('fs');
const path = require('path');
const { pdfToText, docxToText, getOcrPrompt } = require('./lib/gemini_ocr.js');

const GENERAL_DOC_STYLE = `
# CONTEXT: General Document
- **Format**: Standard Japanese document.
- **Line Breaks**: Merge lines within paragraphs.
- **Headings**: Use standard Markdown headings (#, ##, ###) based on the document structure.
`;

async function main() {
    const args = process.argv.slice(2);
    const inputPaths = [];
    let batchSize = 4;
    let startPage = 1;
    let endPage = null;
    let showPrompt = false;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === "--batch_size") batchSize = parseInt(args[++i]);
        else if (args[i] === "--start_page") startPage = parseInt(args[++i]);
        else if (args[i] === "--end_page") endPage = parseInt(args[++i]);
        else if (args[i] === "--show_prompt") showPrompt = true;
        else inputPaths.push(args[i]);
    }

    if (showPrompt) {
        console.log("\n--- Gemini OCR プロンプトテンプレート ---");
        console.log(getOcrPrompt(batchSize, GENERAL_DOC_STYLE));
        console.log("----------------------------------\n");
        return;
    }

    if (inputPaths.length === 0) {
        console.log("-------------------------------------------------------");
        console.log(" PDFファイルまたはフォルダをドロップしてください。");
        console.log(" 使い方: node ocr_general_doc.js <input_path...> [--batch_size <n>]");
        console.log("-------------------------------------------------------");
        return;
    }

    for (const inputPath of inputPaths) {
        const absPath = path.resolve(inputPath);
        if (!fs.existsSync(absPath)) {
            console.error(`[エラー] パスが見つかりません: ${absPath}`);
            continue;
        }

        const processFile = async (filePath) => {
            const ext = path.extname(filePath).toLowerCase();
            if (ext === ".pdf") {
                console.log(`\n[PDF 処理] 開始: ${path.basename(filePath)}`);
                await pdfToText(filePath, batchSize, startPage, endPage, GENERAL_DOC_STYLE);
            } else if (ext === ".docx") {
                console.log(`\n[Word 処理] 開始: ${path.basename(filePath)}`);
                await docxToText(filePath, GENERAL_DOC_STYLE);
            } else {
                console.warn(`[警告] 未対応のファイル形式です: ${path.basename(filePath)}`);
            }
        };

        if (fs.statSync(absPath).isDirectory()) {
            const files = fs.readdirSync(absPath)
                .filter(f => {
                    const ext = f.toLowerCase();
                    return ext.endsWith(".pdf") || ext.endsWith(".docx");
                })
                .sort();
                
            if (files.length === 0) {
                console.warn(`[警告] ディレクトリ内に PDF または Word ファイルが見つかりませんでした: ${absPath}`);
                continue;
            }
            
            console.log(`[情報] ${absPath} 内に ${files.length} 個のファイルが見つかりました`);
            for (let i = 0; i < files.length; i++) {
                const file = files[i];
                const filePath = path.join(absPath, file);
                console.log(`\n[進捗] (${i + 1}/${files.length})`);
                await processFile(filePath);
            }
        } else {
            await processFile(absPath);
        }
    }
    console.log("\nすべての処理が完了しました。");
}

main();
